### PR TITLE
Improving data table documentation

### PR DESCRIPTION
**Pull request by darrenburns**
There were a few features and concepts of the DataTable that were pretty difficult to discover and understand, as they are referenced in docstrings in a way that assumes prior knowledge.

The DataTable widget docs now include a guide at the top, which show off some of the features of the table that I suspect many people don't even realise exist. It also shows how to achieve several things which I see asked regularly.

### CODE REVIEW

File: docs/widgets/data_table.md [hunk 1]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+    ```
+
+### Keys
+
+When adding a row to the table, you can supply a _key_ to [add_row][textual.widgets.DataTable.add_row].
+A key is a unique identifier for that row.
+If you don't supply a key, Textual will generate one for you and return it from `add_row`.
+This key can later be used to reference the row, regardless of its current position in the table.
+
+When working with data from a database for example, you make wish to set the row `key` to the primary key of the data to ensure uniqueness.
```
Comment (willmcgugan) [inline docs/widgets/data_table.md:52]:
```suggestion
When working with data from a database, for example, you may wish to set the row `key` to the primary key of the data to ensure uniqueness.
```

File: docs/widgets/data_table.md [hunk 2]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+The [add_column][textual.widgets.DataTable.add_column] also accepts a `key` argument and works similarly.
+
+Keys are important because cells in a data table can change location due factors like row deletion and sorting.
+Thus, using keys instead of coordinates allows us to refer to data without worrying about its current location in the table.
+
+Sometimes you wish to perform an operation on a cell/row/column based solely on coordinates, without regard for the data that is currently visible at those coordinates.
+In these cases you can make use of the [coordinate_to_cell_key][textual.widgets.DataTable.coordinate_to_cell_key] method to convert
+a coordinate to a _cell key_, which is a `(row_key, column_key)` pair.
```
Comment (willmcgugan) [inline docs/widgets/data_table.md:60]:
```suggestion
If you want to change the table based solely on coordinates, you can use the [coordinate_to_cell_key][textual.widgets.DataTable.coordinate_to_cell_key] method to convert a coordinate to a _cell key_, which is a `(row_key, column_key)` pair.
```

Comment (willmcgugan) [reply docs/widgets/data_table.md:60]: Originally felt a little too wordy...

File: docs/widgets/data_table.md [hunk 3]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+
+### Cursors
+
+You can retrieve the current coordinate of the cursor from the `cursor_coordinate` reactive attribute.
```
Comment (willmcgugan) [inline docs/widgets/data_table.md:64]:
```suggestion
The coordinate of the cursor is exposed via the `cursor_coordinate` reactive attribute.
```

Original wording suggests to me that there may be some additional step involved.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Great. Just a few suggestions.

I'm sure @rodrigogiraoserrao will pick up on the typos.

File: docs/widgets/data_table.md [hunk 4]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
     --8<-- "docs/examples/widgets/data_table.py"
     ```
 
+To add rows and columns one-at-a-time, use [add_row][textual.widgets.DataTable.add_row] and [add_column][textual.widgets.DataTable.add_column].
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:27]: “one-at-a-time” feels awkward to me.
```suggestion
To add a single row or column use [add_row][textual.widgets.DataTable.add_row] and [add_column][textual.widgets.DataTable.add_column], respectively.
```

File: docs/widgets/data_table.md [hunk 5]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+
+#### Styling and justifying cells
+
+Cells can contain more than just plain strings - [Rich](https://rich.readthedocs.io/en/stable/introduction.html) renderables such as [`Text`](https://rich.readthedocs.io/en/stable/text.html?highlight=Text#rich-text) are also supported.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:31]:
```suggestion
Cells can contain more than just plain strings - [Rich][rich] renderables such as [`Text`][rich.text.Text] are also supported.
```

File: docs/widgets/data_table.md [hunk 6]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+
+### Keys
+
+When adding a row to the table, you can supply a _key_ to [add_row][textual.widgets.DataTable.add_row].
+A key is a unique identifier for that row.
+If you don't supply a key, Textual will generate one for you and return it from `add_row`.
+This key can later be used to reference the row, regardless of its current position in the table.
+
+When working with data from a database for example, you make wish to set the row `key` to the primary key of the data to ensure uniqueness.
+The [add_column][textual.widgets.DataTable.add_column] also accepts a `key` argument and works similarly.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:53]:
```suggestion
The method [add_column][textual.widgets.DataTable.add_column] also accepts a `key` argument and works similarly.
```

File: docs/widgets/data_table.md [hunk 7]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+
+Keys are important because cells in a data table can change location due factors like row deletion and sorting.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:55]:
```suggestion
Keys are important because cells in a data table can change location due to factors like row deletion and sorting.
```

File: docs/widgets/data_table.md [hunk 8]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+Thus, using keys instead of coordinates allows us to refer to data without worrying about its current location in the table.
+
+Sometimes you wish to perform an operation on a cell/row/column based solely on coordinates, without regard for the data that is currently visible at those coordinates.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md]:
```suggestion
Sometimes you may wish to perform an operation on a cell/row/column based solely on coordinates, without regard for the data that is currently visible at those coordinates.
```

File: docs/widgets/data_table.md [hunk 9]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+In these cases you can make use of the [coordinate_to_cell_key][textual.widgets.DataTable.coordinate_to_cell_key] method to convert
+a coordinate to a _cell key_, which is a `(row_key, column_key)` pair.
+
+### Cursors
+
+You can retrieve the current coordinate of the cursor from the `cursor_coordinate` reactive attribute.
+Three types of cursor are supported: `cell`, `row`, and `column`.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:63]:
```suggestion
Three types of cursors are supported: `cell`, `row`, and `column`.
```

File: docs/widgets/data_table.md [hunk 10]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+    ```{.textual path="docs/examples/widgets/data_table_fixed.py"}
+    ```
+
+=== "data_table_fixed.py"
+
+    ```python
+    --8<-- "docs/examples/widgets/data_table_fixed.py"
+    ```
+
+In the example above, we set `fixed_rows` to `2`, and `fixed_columns` to `1`,
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:126]: Would be brilliant if the example also showed this. Maybe add 100 dummy rows to the table and then scroll to the bottom?

File: docs/widgets/data_table.md [hunk 11]
```diff
@@ -20,11 +24,143 @@ The example below populates a table with CSV data.
...
+
+The `DataTable` can be sorted using the [sort][textual.widgets.DataTable.sort] method.
+In order to sort your data by a column, you must have supplied a `key` to the `add_column` method
+when you added it.
+You can then pass this key to the `sort` method to sort by that column.
+Additionally, you can sort by multiple columns by passing multiple keys to `sort`.
+
+### Labelled rows
+
+A "label" can be attached to rows using the [add_row][textual.widgets.DataTable.add_row] method.
```
Comment (rodrigogiraoserrao) [inline docs/widgets/data_table.md:140]:
```suggestion
A "label" can be attached to a row using the [add_row][textual.widgets.DataTable.add_row] method.
```

File: docs/widgets/data_table.md (hunk 5 above)
Comment (darrenburns) [reply docs/widgets/data_table.md:31]: What effect does this have?

File: docs/widgets/data_table.md (hunk 8 above)
Comment (darrenburns) [reply docs/widgets/data_table.md]: Outdated now

File: docs/widgets/data_table.md (hunk 10 above)
Comment (darrenburns) [reply docs/widgets/data_table.md:126]: Hmm, good idea! Will do that shortly.

File: docs/widgets/data_table.md (hunk 5 above)
Comment (rodrigogiraoserrao) [reply docs/widgets/data_table.md:31]: Ah, sorry for my oversight!
